go/vt/vtctl/reparentutil/emergency_reparenter.go

```diff
@@ -642,71 +640,32 @@ func (erp *EmergencyReparenter) identifyPrimaryCandidate(
-	// Are there any replicas with better promotion rules?
-	// Maybe we actually promoted such a replica. Does that mean we should keep it?
-	// Maybe we promoted a "neutral", and some "prefer" server is available.
-	// Maybe we promoted a "prefer_not"
-	// Maybe we promoted a server in a different cell than the primary
-	// There's many options. We may wish to replace the server we promoted with a better one.
-
-	// If the user requested for prevention of cross cell promotion then we should only search for valid candidates in the same cell
-	// otherwise we can search in any cell
-	if opts.PreventCrossCellPromotion {
-		// find candidates in the same cell from the preferred candidates list
-		candidate = findCandidateSameCell(intermediateSource, prevPrimary, preferredCandidates)
-		if candidate != nil {
-			return candidate, nil
-		}
-		// we do not have a preferred candidate in the same cell
-	} else {
-		// find candidates in any cell from the preferred candidates list
-		candidate = findCandidateAnyCell(intermediateSource, preferredCandidates)
-		if candidate != nil {
-			return candidate, nil
-		}
-	}
 
-	// repeat the same process for the neutral candidates list
-	if opts.PreventCrossCellPromotion {
-		// find candidates in the same cell from the neutral candidates list
-		candidate = findCandidateSameCell(intermediateSource, prevPrimary, neutralReplicas)
-		if candidate != nil {
-			return candidate, nil
-		}
-		// we do not have a neutral candidate in the same cell
-	} else {
-		// find candidates in any cell from the neutral candidates list
-		candidate = findCandidateAnyCell(intermediateSource, neutralReplicas)
+	// We have already selected an intermediate source which was selected based on the replication position
+	// (ties broken by promotion rules), but that tablet might not even be a valid candidate i.e. it could
+	// be in a different cell when we have PreventCrossCellPromotion specified, or it could have a promotion rule of
+	// MustNot. Even if it is valid, there could be a tablet with a better promotion rule. This is what we try to
+	// find here.
+	// We go over all the promotion rules in descending order of priority and try and find a valid candidate with
+	// that promotion rule.
+	// If the intermediate source has the same promotion rules as some other tablets, then we prioritize using
+	// the intermediate source since we won't have to wait for the new candidate to catch up!
+	for _, promotionRule := range promotionrule.AllPromotionRules() {
+		candidates := getTabletsWithPromotionRules(validCandidates, promotionRule)
+		candidate = findCandidate(intermediateSource, candidates)
 		if candidate != nil {
 			return candidate, nil
 		}
 	}
-
-	// return any valid tablet
-	return findCandidateAnyCell(intermediateSource, validCandidates), nil
+	// Unreachable
+	// We should have found atleast 1 tablet in the valid list
+	// If the list is empty then we should have errored out much sooner
```

Comment:
punctuate to break up these thoughts please :)

Reply:
Fixed